File: src/term.ts
// Nothing this vault prints is text it wrote. A path, a title, a link target
// come from a note a human or an LLM authored; an error message now quotes up
// to 200 bytes of whatever answered the embedding endpoint. A bare `\r` or an
// ESC sequence in any of them rewrites the line the terminal has already drawn
// — so it goes through here first, on stdout and stderr alike.

/** Untrusted text, safe to echo: control characters print as `?`. */
export const safe = (text: string): string => text.replace(/\p{Cc}/gu, "?");

/**
 * An untrusted error as the message it was written as — never a stack trace,
 * never a redrawn terminal. Scrubbed line by line, so the newlines an error
 * deliberately carries (a usage message under a bad flag) still break lines.
 */
export const printable = (e: unknown): string =>
  (e instanceof Error ? e.message : String(e)).split("\n").map(safe).join("\n");

File: src/wilcus_vault/term.py
"""Text safety for anything the vault prints or raises."""

import re

# Unicode category Cc: C0 and C1 control characters. A bare CR or an ESC
# sequence inside a note title would otherwise rewrite the terminal line.
_CONTROL = re.compile(r"[\x00-\x1f\x7f-\x9f]")


class VaultError(Exception):
    """An error the vault raised on purpose, with a message meant for a human."""


def safe(text: str, keep: str = "") -> str:
    """Untrusted text, safe to echo: control characters print as `?`, except any in `keep`."""
    return _CONTROL.sub(lambda m: m[0] if m[0] in keep else "?", text)


def printable(error: object) -> str:
    """An error as its message, scrubbed line by line so deliberate newlines survive."""
    return "\n".join(safe(line) for line in str(error).split("\n"))
